Support for Kestrel server using multiple encodings.

samuelhewitt opened the pull request:
### Summary

When using Kestrel server and handling a request with multiple encodings, the middleware is failing to recognize the encodings and decode. eg. `Content-Encoding: gzip, br`

### Changes

- When ContentEncoding StringValues has a single entry, check for a comma. If present, split on commas and proceed.
- Add unit tests

### Outcome

- Kestrel Server - `Content-Encoding: gzip, br` - successfully decoded

### Additional Information

- Http1Connection loads headers here:
https://github.com/dotnet/aspnetcore/blob/5aa86873c30f1f4643c801bd303901fb66af8b09/src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.Generated.cs#L6527-L6535

 - Since there is only a single Content-Encoding header, new StringValues(valueStr) is called where valueStr == "gzip, br"
https://github.com/dotnet/aspnetcore/blob/5aa86873c30f1f4643c801bd303901fb66af8b09/src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.Generated.cs#L6630

 - When using TryGetValue, a single entry is retrieved
https://github.com/dotnet/aspnetcore/blob/5aa86873c30f1f4643c801bd303901fb66af8b09/src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.Generated.cs#L1876-L1880

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/alexanderkozlenko/aspnetcore-request-decompression?pullRequest=111) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/alexanderkozlenko/aspnetcore-request-decompression?pullRequest=111) before we can accept your contribution.<br/><hr/>**shewitt** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/alexanderkozlenko/aspnetcore-request-decompression?pullRequest=111) it.</sub>

alexanderkozlenko commented:
Is it worth to use `String.Split(char[]?, StringSplitOptions)` with static array of  separators (`'  '`  and `','` ) instead of `String.Replace(string, string?, StringComparison)` with `String.Split(params char[]?)`? According to my rought benchmarks it should be a bit faster:

```
| Method |      Mean |    Error |   StdDev | Ratio |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |----------:|---------:|---------:|------:|-------:|------:|------:|----------:|
|   O1E2 | 135.98 ns | 2.653 ns | 2.949 ns |  1.00 | 0.0203 |     - |     - |     160 B |
|   O2E2 |  85.71 ns | 1.709 ns | 1.755 ns |  0.63 | 0.0202 |     - |     - |     160 B |
```

samuelhewitt commented:
> Is it worth using `String.Split(char[], StringSplitOptions)` with static array of separators (`' '`, `','` ) instead of `String.Replace(string, string, StringComparison)` with `String.Split(params char[])`? According to my rought benchmarks it should be a bit faster:

Thinking about my approach and this one, I realize that neither would allow for an encoding that has a space in the name (`gzip` vs `my encoding`). Is that something that is allowed by the spec?

If so, the code will need to split on commas and then trim each entry. If its not something that needs to be handled, happy to switch to your approach.

alexanderkozlenko commented:
As far as I see from ABNFs listed here, the space character is not allowed:

- [RFC 7231, Appendix D "Collected ABNF"](https://tools.ietf.org/html/rfc7231#appendix-D)
- [RFC 7230, Section 3.2.6 "Field Value Components"](https://tools.ietf.org/html/rfc7230#section-3.2.6)

samuelhewitt commented:
Interesting - after reading that, there must be a built in function to support the various restricted characters and quoting rules - which I was able to find.
I've updated the PR to use the `.GetCommaSeparatedValues` extension method. Let me know what you think.

alexanderkozlenko commented on src/Anemonis.AspNetCore.RequestDecompression/RequestDecompressionMiddleware.cs:
```diff
@@ -93,7 +93,7 @@ public async Task InvokeAsync(HttpContext context, RequestDelegate next)
             }
 
-            if (!context.Request.Headers.TryGetValue(HeaderNames.ContentEncoding, out var encodingNames) && (encodingNames.Count > 0))
+            if (!context.Request.Headers.ContainsKey(HeaderNames.ContentEncoding))
```
I suppose this check could be omitted since [`GetCommaSeparatedValues`](https://docs.microsoft.com/ru-ru/dotnet/api/microsoft.aspnetcore.http.headerdictionaryextensions.getcommaseparatedvalues?view=aspnetcore-3.1) will return an empty array if `Content-Encoding` header is not present.

samuelhewitt replied:
I left this check to have a fast-path on requests without ContentEncoding present. Although not too expensive, it looks like GetCommaSeparateValues at minimum will allocate a few int's as part of the HeaderSegmentCollection, even when the string is empty.

alexanderkozlenko commented on src/Anemonis.AspNetCore.RequestDecompression/RequestDecompressionMiddleware.cs:
```diff
@@ -109,12 +109,23 @@ public async Task InvokeAsync(HttpContext context, RequestDelegate next)
+            // There could be a single StringValues entry with comma delimited contents
+            //  Content-Encoding: gzip, br, "someEncoding"
+            //  string[] { "gzip", "br", "someEncoding"}
+            var encodingNamesParsed = context.Request.Headers.GetCommaSeparatedValues(HeaderNames.ContentEncoding);
```
Could you please consider reusing the original name `encodingNames`?

samuelhewitt replied:
Done.

alexanderkozlenko commented:
> Interesting - after reading that, there must be a built in function to support the various restricted characters and quoting rules - which I was able to find.
> I've updated the PR to use the `.GetCommaSeparatedValues` extension method. Let me know what you think.

Looks great!

alexanderkozlenko commented:
Version 1.8.0 with these changes has been published to NuGet. Thank you for your contribution!
